File: src/main/java/nl/mpcjanssen/simpletask/sort/ProjectComparator.java
package nl.mpcjanssen.simpletask.sort;

import com.google.common.collect.Ordering;
import nl.mpcjanssen.simpletask.task.Task;

import java.util.Collections;
import java.util.List;

public class ProjectComparator extends Ordering<Task> {

    private final AlphabeticalStringComparator mStringComparator;

    public ProjectComparator (boolean caseSensitive) {
        super();
        this.mStringComparator = new AlphabeticalStringComparator(caseSensitive);
    }


    @Override
    public int compare(Task a, Task b) {
        if (a == b) {
            return 0;
        } else if (a == null) {
            return -1;
        } else if (b == null) {
            return 1;
        }
        List<String> projectsA = a.getTags();
        List<String> projectsB = b.getTags();

        if (projectsA.isEmpty() && projectsB.isEmpty()) {
            return 0;
        } else if (projectsA.isEmpty()) {
            return 1;
        } else if (projectsB.isEmpty()) {
            return -1;
        } else {
            Collections.sort(projectsA);
            Collections.sort(projectsB);
            return mStringComparator.compare(projectsA.get(0),projectsB.get(0));
        }
    }
}

File: src/main/java/nl/mpcjanssen/simpletask/sort/ProjectComparator.kt
package nl.mpcjanssen.simpletask.sort

import nl.mpcjanssen.simpletask.task.Task
import java.util.*

class ProjectComparator(caseSensitive: Boolean) : Comparator<Task> {

    private val mStringComparator: AlphabeticalStringComparator

    init {
        this.mStringComparator = AlphabeticalStringComparator(caseSensitive)
    }


    override fun compare(a: Task?, b: Task?): Int {
        if (a === b) {
            return 0
        } else if (a == null) {
            return -1
        } else if (b == null) {
            return 1
        }
        val projectsA = a.tags.toMutableList()
        val projectsB = b.tags.toMutableList()

        if (projectsA.isEmpty() && projectsB.isEmpty()) {
            return 0
        } else if (projectsA.isEmpty()) {
            return 1
        } else if (projectsB.isEmpty()) {
            return -1
        } else {
            Collections.sort(projectsA)
            Collections.sort(projectsB)
            return mStringComparator.compare(projectsA[0], projectsB[0])
        }
    }
}
